index.d.ts

```diff
@@ -1,288 +1,591 @@
-declare module 'react-table' {
-  import { ReactNode, useState } from 'react'
+// TypeScript Version: 3.5
+
+import { ReactNode, useState, ComponentType, MouseEvent } from 'react'
+
+/**
+ * The empty definitions of below provides a base definition for the parts used by useTable, that can then be extended in the users code.
+ *
+ * @example
+ *  export interface TableOptions<D extends object = {}}>
+ *    extends
+ *      UseExpandedOptions<D>,
+ *      UseFiltersOptions<D> {}
+ */
+export interface TableOptions<D extends object> extends UseTableOptions<D> {}
+
+export interface TableInstance<D extends object = {}>
+  extends Omit<TableOptions<D>, 'columns'>,
+    UseTableInstanceProps<D> {}
+
+export interface TableState<D extends object = {}> {} /* tslint:disable-line no-empty-interface */ // eslint-disable-line @typescript-eslint/no-empty-interface
+
+export interface Hooks<D extends object = {}> extends UseTableHooks<D> {}
+
+export interface Cell<D extends object = {}> extends UseTableCellProps<D> {}
+
+export interface Column<D extends object = {}>
+  extends UseTableColumnOptions<D> {}
+
+export interface ColumnInstance<D extends object = {}>
+  extends Omit<Column<D>, 'id'>,
+    UseTableColumnProps<D> {}
+
+export interface HeaderGroup<D extends object = {}>
+  extends ColumnInstance<D>,
+    UseTableHeaderGroupProps<D> {}
+
+export interface Row<D extends object = {}> extends UseTableRowProps<D> {}
+
+/* #region useTable */
+export function useTable<D extends object = {}>(
+  options: TableOptions<D>,
+  ...plugins: PluginHook<D>[]
+): TableInstance<D>
+
+export type UseTableOptions<D extends object> = Partial<{
+  columns: Column<D>[]
+  state: TableStateTuple<D>
+  data: D[]
+  defaultColumn: Partial<Column<D>>
+  initialRowStateKey: IdType<D>
+  getSubRows: (row: Row<D>, relativeIndex: number) => Row<D>[]
+  getRowID: (row: Row<D>, relativeIndex: number) => string
+  debug: boolean
+}> &
+  Record<string, any>
```

Comment:
I'm not a big fan of the `Record` here, it effectively means that you can pass any value in to this with no easy way to tell that it's junk, and it negates the ability to extend `TableOptions`

Reply:
So let people extend TableOptions by creating an interface that lists the extra fields that they want to add and typecheck them.

Once you add Record here, there's no way that someone who wants TableOptions to be typechecked to undo the impact of such a change.  I don't know if you can remove it again after the fact.

Reply:
BTW of all of the comments that I've made, this might be the one that is most important to me, the others can be fixed or modified by overrides or casts, I can't see how to undo this one if it's published like this.  I can see how it might be useful to folks, but there are other ways to achieve this effect.

Reply:
Well, at the moment, none of the typings are published. So until we do, this is revert-able.

Reply:
that could work, though that extra param would filter up the tree a bit.

Reply:
Unfortunately, my suggestion won't work. We'll have to revisit this

Reply:
I should add that removing the Record here exposes the Partial nature of TableOptions in TableInstance and you start running into issues where you can't destructure state inline because it might be undefined -which I think that it can't be.


I think that it might be more accurate to define TableInstance as

```tsx
export interface TableInstance<D extends object = {}>
  extends Required<Omit<TableOptions<D>, 'columns'>>,
    UseTableInstanceProps<D> {}
```

Reply:
Alternatively, moving the `Partial` up to where it is used in the options.

Reply:
The only reason that I can come up with re. choosing between adding Required or moving Parial is that if we add Required then the Required/Optional code won't need to be repeated when extending TableInstance, if we move the Partial (in more than just UseTableOptions, which we might for consistency), then that would need to be replicated when extending the interfaces for other plugins.